plugins/inputs/mqtt_consumer/mqtt_consumer.go

```diff
@@ -279,26 +259,72 @@ func (m *MQTTConsumer) recvMessage(_ mqtt.Client, msg mqtt.Message) {
 	}
 }
 
+// compareTopics is used to support the mqtt wild card `+` which allows for one topic of any value
+func compareTopics(expected string, incoming string) bool {
+	expectedSplit := strings.Split(expected, "/")
+	incomingSplit := strings.Split(incoming, "/")
+	if len(expectedSplit) != len(incomingSplit) {
+		return false
+	}
+
+	for i := range expectedSplit {
+		if incomingSplit[i] != expectedSplit[i] && expectedSplit[i] != "+" {
+			return false
+		}
+	}
+
+	return true
+}
+
 func (m *MQTTConsumer) onMessage(acc telegraf.TrackingAccumulator, msg mqtt.Message) error {
+
 	metrics, err := m.parser.Parse(msg.Payload())
 	if err != nil {
 		return err
 	}
 
-	if m.chosenTopicTag != "" {
-		topic := msg.Topic()
-		for _, metric := range metrics {
-			metric.AddTag(m.chosenTopicTag, topic)
+	for _, metric := range metrics {
+		if m.topicTagParse != "" {
+			metric.AddTag(m.topicTagParse, msg.Topic())
+		}
+		for _, p := range m.TopicParsing {
+			if compareTopics(p.Topic, msg.Topic()) {
+				values := strings.Split(msg.Topic(), "/")
+
+				if p.Measurement != "" {
+					m, err := parseMeasurement(strings.Split(p.Measurement, "/"), values)
+					if err != nil {
+						return err
+					}
+					metric.SetName(m)
+				}
+				if p.Tags != "" {
+					tags, err := parseTags(strings.Split(p.Tags, "/"), values)
+					if err != nil {
+						return err
+					}
+					for k, v := range tags {
+						metric.AddTag(k, v)
+					}
+				}
+				if p.Fields != "" {
+					fields, err := parseFields(strings.Split(p.Fields, "/"), values, p.FieldTypes)
+					if err != nil {
+						return err
+					}
+					for k, v := range fields {
+						metric.AddField(k, v)
+					}
+				}
+			}
```

Comment:
Save one indentation level
```suggestion
			if !compareTopics(p.Topic, msg.Topic()) {
				continue
			}

			values := strings.Split(msg.Topic(), "/")
			if p.Measurement != "" {
				m, err := parseMeasurement(strings.Split(p.Measurement, "/"), values)
				if err != nil {
					return err
				}
				metric.SetName(m)
			}
			if p.Tags != "" {
				tags, err := parseTags(strings.Split(p.Tags, "/"), values)
				if err != nil {
					return err
				}
				for k, v := range tags {
					metric.AddTag(k, v)
				}
			}
			if p.Fields != "" {
				fields, err := parseFields(strings.Split(p.Fields, "/"), values, p.FieldTypes)
				if err != nil {
					return err
				}
				for k, v := range fields {
					metric.AddField(k, v)
				}
			}
```